### DISCUSSION TITLE
Adding “text styling toolbar” to Add to cart button

### DISCUSSION THREAD

**Discussion by imanish003** [woocommerce/woocommerce#44972] [WooCommerce, Blocks, and Site Editing]
## Goal of this Discussion?

I have a few questions that I posed at the end of this post. I would greatly appreciate it if you could share your answers, thoughts, or opinions in the comments section.🙂

## Introduction

I wanna add following toolbar controls to `Add to cart` button toolbar, which is also available on `core/button`.
![image](https://user-images.githubusercontent.com/16707866/227517519-fa40c8aa-c95f-4128-8382-e676297ff365.png)
I will call this toolbar, `text styling toolbar` in this post.

## How it's implemented in `core/button`?

These are visible on core button because it uses `RichText` component.

RichText Doc: [https://github.com/WordPress/gutenberg/blob/HEAD/packages/block-editor/src/components/rich-text/README.md](https://github.com/WordPress/gutenberg/blob/HEAD/packages/block-editor/src/components/rich-text/README.md)

## Can I use `RichText` for Add to cart button?

We do not want to allow users to style specific parts of the button text. The entire button text should be styled when using the styling toolbar because we show `product(add to cart)` button text based on product type & stock status. Therefore, we can't allow the styling of partial text of button.

For those who are not familiar with `RichText`, it's used by many components. For example, `heading` component in Gutenberg. Here is a quick demo

https://user-images.githubusercontent.com/16707866/227518300-37ea8e33-094d-4490-b598-4aa78e03ea6a.mov

## Why we can't allow users to style the partial text of Add to cart button

We show Add to cart button text based on product type & stock status. For example, If the product type is *variable* then we will need to show `Select options` text as shown in screenshot below
![image](https://user-images.githubusercontent.com/16707866/227517960-d0fa7b49-28fe-4912-913b-f7491144df86.png)

Now let's say we allow users to style any specific part of button text. If user makes following changes to `Add to cart` text:

- Make "Add" text bold
- Make "to cart" italic

Now how should text look for other buttons, i.e., Select options, View products? I hope you got the point that we can't allow users to style parts of button text.

## Workarounds for Preventing Partial Text Styling in the RichText Component

There isn’t a direct way to achieve this in RichText. Some of the hacky ways could be:

1. We can use **`window.getSelection()`** API to get currently selected text. We will only show the text styling toolbar if all the button text is selected. This way user won’t be able to style specific parts of button text. **Cons:** I think UX won’t be great because a user might not realize that he must select all the text of button to style it.
2. Whenever a user styles partial text, we can discard the styling. This can be done in the `onChange` prop of RichText. **Cons:** I believe that in this case, the user experience won't be great because when a user selects any part of the button text, the text styling toolbar will be displayed. This will give the wrong impression that the user can style that text, when in reality we will be discarding those changes.
3. Whenever a user styles any part of the text, we apply the styling to the entire text. **Cons:** If a user selects a part of the text and applies styling, they might expect the styling to be applied only to that text. However, this is not the case, as we apply the styling to all of the button text.

## Questions:
- Do we even need to add a "text styling toolbar" if it requires us to use a hacky way to achieve this? We can achieve the same effect using the Typography setting available in the sidebar.
![image](https://user-images.githubusercontent.com/16707866/227518950-74bec2f9-3e77-41ce-b894-7bb14a3e91f6.png)
- If we still want to add a "text styling toolbar" for the "Add to cart" button, what hacky approach do you think would be best? In my opinion, the third approach would be better for our case.

**Comment by roykho** [comment]
Thanks for putting this together.

> We do not want to allow users to style specific parts of the button text.

I understand the `add to cart` text will be different depending on if the product is a "simple" or "variable" and also the stock status. But that's ok right? The styling would just apply to any text that is displayed, no?

Or perhaps another idea is to add a button in the "text styling toolbar" at the bottom that says `Apply to all product buttons`. So the user can choose if they actually want to have all buttons look the same.

**Reply by imanish003** [reply]
> I understand the add to cart text will be different depending on if the product is a "simple" or "variable" and also the stock status. But that's ok right? The styling would just apply to any text that is displayed, no?

Although I tried to explain it in "Why we can't allow users to style the partial text of Add to cart button" section but looks like I was not able to explain it clearly. Let me try again 🙂

If we allow users to style specific parts of the button text, then there could be many cases, some of which are:
### Case 1: All the text is bold
<img width="172" alt="image" src="https://user-images.githubusercontent.com/16707866/227891891-3900171e-39c1-478e-b803-aa4c13e27a78.png">

![image](https://user-images.githubusercontent.com/16707866/227891213-13671c34-f2e5-4edb-8672-c520e36d9055.png)

### Case 2: Mix of bold & italic for words
<img width="168" alt="image" src="https://user-images.githubusercontent.com/16707866/227891982-e1aa0308-1ddc-43ab-876b-48cc9fa57b8a.png">

![image](https://user-images.githubusercontent.com/16707866/227891291-5bcf83a4-b6aa-45f6-a351-de93060e77ec.png)
As you can see in above screenshot,

- _Add_ is bold
- _to_ is bold as well as italic
- _cart_ is bold

## Let's understand using _Case 2_

As we know, text can vary depending on the product type and its availability. Here are some common values:
- "Add to cart": This is the default text for simple products that are in stock and available for purchase.
- "Select options": For variable products, this text prompts the user to choose product variations (e.g., size, color) before adding the item to their cart.
- "Read more": This text may be displayed for products that are not available for direct purchase, like external/affiliate products or products without a price set.
- "View products": This is typically used for grouped products, which are a collection of related products that can be purchased individually.
- "Out of stock": If a product is not in stock, this text may be displayed instead of the "Add to cart" button.

Now the problem is, how do we deterministically translate the styling set by user in Case 2 to all these different texts? For example, if button text is "Select options".  
- Should "Select" text styling be bold or italic, or combination of both?
- Should "Options" text styling be bold or italic or combination of both?

We can't answer this question deterministically. I hope I was able to explain what I am thinking. I know it's a bit complicated. Please let me know if you have any further questions. 🙂

**Reply by roykho** [reply]
Ah ok so I think I see what you mean. So the style changes are NOT only CSS but actual HTML gets applied such as `<strong>Select</strong> Options` and there is not way for us to know which word/text to apply that if the product is later changed to `simple product` which has `Add to Cart` text for example. Is that what you mean?

**Reply by imanish003** [reply]
Yes, that's correct. 🙂

**Comment by tjcafferkey** [comment]
Thanks for this suggestion @imanish003 here are my thoughts

>Do we even need to add a "text styling toolbar" if it requires us to use a hacky way to achieve this? We can achieve the same effect using the Typography setting available in the sidebar.

We should try and align with Gutenbergs behaviour as much as possible, I'd be interested to see in what blocks do they have the "text styling toolbar" options inside the settings sidebar of the block, like in the screenshot you've mentioned above. From my point of view this is confusing for the user as they have the consider two different ways to achieve the same thing. If Gutenberg provides a standard way for us to give the user the ability to add text decoration like with the "text styling toolbar" then we should be using it?

Something to run past Shaun for confirmation perhaps if you can TLDR this conversation for him.

>If we still want to add a "text styling toolbar" for the "Add to cart" button, what hacky approach do you think would be best? In my opinion, the third approach would be better for our case.

Here is what I think, and it sounds like we are aligned.

>Whenever a user styles any part of the text, we apply the styling to the entire text. Cons: If a user selects a part of the text and applies styling, they might expect the styling to be applied only to that text. However, this is not the case, as we apply the styling to all of the button text.

This is what I would expect as something similar already happens in the following scenario:

1. Add paragraph block, and some text
2. Select part of the text and change the color of this
3. The color is applied to all the text in the block.

**Reply by imanish003** [reply]
Hi @tjcafferkey, it's great to hear that you agree that the third approach is the better one if we decide to implement the "text styling toolbar" 🙌🏻

> We should try and align with Gutenbergs behaviour as much as possible, I'd be interested to see in what blocks do they have the "text styling toolbar" options inside the settings sidebar of the block, like in the screenshot you've mentioned above. From my point of view this is confusing for the user as they have the consider two different ways to achieve the same thing. If Gutenberg provides a standard way for us to give the user the ability to add text decoration like with the "text styling toolbar" then we should be using it?

Actually, Gutenberg also provides two ways. Here is a quick demo with Paragraph block:

https://user-images.githubusercontent.com/16707866/228748283-fa9fcb00-8243-4046-92da-b89a5fbd831c.mov

The same could be seen for `core/button` too.

Although following options are not available in sidebar, which are available in "text styling toolbar":
- Highlight
- Inline code
- Inline image
- Keyboard input 
- Subscript
- Superscript

In my opinion, these options don't make sense for the "Add to cart" button. The typography options available in the sidebar are sufficient. Perhaps we could consider adding a "text styling toolbar" in the future if there is sufficient demand from the WooCommerce community. What do you think?

**Reply by tjcafferkey** [reply]
>Perhaps we could consider adding a "text styling toolbar" in the future if there is sufficient demand from the WooCommerce community. What do you think?

Sorry @imanish003 I left a reaction to your latest comment but not a comment to confirm I agree with this latest statement. I agree in that I think the available typography options from the sidebar are sufficient, and that unless there's demand or direction from Shaun to implement the _text styling toolbar_ then we can descope this.

**Reply by imanish003** [reply]
No worries! Thanks for confirming, @tjcafferkey. For now, I will go ahead and will descope the task. We can always revisit this in the future if we need to. 🤝
